### PR TITLE
Fix #82: make swallowed IO errors fail toward affected instead of under-building

### LINKED ISSUE #82: [bug] Swallowed IO errors bias the analysis toward under-building (the unsafe direction)
## Behavior

Several catch blocks in the analysis path bias the result toward **under-building**, which for a tool whose worst failure mode is a green build on untested code is the unsafe direction. The same files apply the conservative (over-build) posture elsewhere, so this is inconsistent as well as risky.

Sites (checked against current main):

- `PomChangeAnalyzer` `checkFileForPropertyRefs`: catches `IOException` with a "skip unreadable files" comment and no logging. An unreadable filtered resource is treated as NOT containing the changed property, so the module is not marked affected. Two branches earlier, an oversized file returns *affected*: opposite risk postures in the same method.
- `scanDirectoryForPropertyRefs`: skips an unreadable directory with no logging.
- `isBinaryFile`: I/O failure swallowed with no logging.
- `readPomText`: returns null on failure with only a debug log; the caller then silently skips the whole "child references changed property" check.
- `getChangedTransitiveDependencyScope` (participant): a `DependencyResolutionException` is logged at debug with no partial result, and the module is treated as unaffected by a changed managed dependency.

## Expected

Each site should either mark the module affected or at least log at WARN, never silently under-build. `PomChangeAnalyzer` already has the right template in its catch-all that conservatively marks all dependents affected.

A per-build counter of swallowed I/O errors, surfaced in the summary log line and the JSON report, would let CI alert on degraded analysis.

## Suggested fix

I have a branch with a failing-test-first implementation covering six sites (unreadable resource, unreadable directory, resolution failure, etc., all asserting the conservative outcome). Happy to rebase and PR it if the direction is welcome.

### PR DESCRIPTION
**Pull request by paoloantinori**
## Problem

As filed in #82: several catch blocks biased the analysis toward under-building, the unsafe direction for a tool whose worst failure mode is a green build on untested code. An unreadable filtered resource was treated as NOT containing the changed property (module not affected) while an oversized file two branches earlier returned affected; an unreadable directory was skipped silently; a `DependencyResolutionException` was logged at debug and the module treated as unaffected by a changed managed dependency.

## Change

Six sites aligned to the conservative posture (the same posture the codebase already applies in its broad catch-all in `analyzeChanges`):

- `checkFileForPropertyRefs` IOException -> WARN + return true (affected)
- `scanDirectoryForPropertyRefs` IOException -> WARN + return true
- `isBinaryFile` IOException -> WARN added (return already conservative)
- `readPomText` failure -> WARN; the caller marks the child affected instead of silently skipping the child-references check
- `getChangedTransitiveDependencyScope` null result -> returns `"compile"`, the most impactful scope, so both consumers (skip-tests assignment and transitive-affected) treat the module conservatively
- POSIX permission tests guarded for root-in-Docker environments

## Deterministic proof (Byteman fault injection, optional profile)

Includes a `-Pbyteman` profile (test-scope `org.jboss.byteman:byteman:4.0.27`, zero impact on the default build): a rule makes the filtered-resource file unreadable exactly at the check entry, so the real code throws the real IOException into the real catch, on any OS and in root containers where chmod-based tests are vacuous. Same test, two outcomes: **red on unmodified main** (module silently skipped, the exact under-build), **green with this fix**. The test carries a fired-flag health assertion so a rule that fails to arm fails the test instead of passing vacuously.

## Verification

RED commit first (tests failing on unmodified main), then the fix. Full `core`+`extension3` suites green; byteman twin test red-on-main / green-on-fix verified on both sides.

## Summary by CodeRabbit

- **Bug Fixes**
  - Modules are now conservatively marked as affected when dependency resolution, model generation, or resource analysis fails.
  - Improved warnings provide clearer visibility when affected status cannot be determined reliably.
  - Large, unreadable, or invalid project resources no longer cause potentially affected modules to be skipped.

- **Reporting**
  - Added the `TRANSITIVE_DEPENDENCY_UNRESOLVED` reason to affected-module reports and validation schemas.

- **Documentation**
  - Updated the report format documentation to describe the new affected-module reason.

### CODE REVIEW

**Comment by paoloantinori** [comment]
One known minor edge, flagged during self-review for the record: in `getChangedTransitiveDependencyScope`, the conservative `return "compile"` on a resolution failure fires before the collectCache is populated, so the two consumers (transitive-affected and skip-tests) can see different outcomes if a retry succeeds on the second call. The direction of any inconsistency is still toward over-building rather than under-building, and it matches pre-existing semantics, so I left it alone; caching the failure would be a small follow-up if you consider the flaky-resolution case worth hardening.

**Review by gnodet — APPROVED** [review]
Well-scoped bug fix that correctly aligns six IO error handling sites with the established conservative pattern (mark affected on failure, never silently under-build). TDD approach with RED-first commits and thoughtful POSIX-permission guards demonstrates strong engineering discipline.

**Minor observations (non-blocking):**

1. **Orphaned byteman rule file** — `extension3/src/test/resources/byteman/fault-injection.btm` was added in the spike commit but the byteman harness (dependency, profile, Java test) was removed in a later commit. This `.btm` file is now orphaned and could be deleted to avoid confusion.

2. **Missing test for `getChangedTransitiveDependencyScope` null-result path** — The fix returning `"compile"` instead of `null` when `DependencyResolutionException` has no partial result lacks a dedicated test. The existing `ScalpelLifecycleParticipantTest` already mocks this exception with a non-null partial result, so adding a sibling test with a null-result exception would be straightforward.

### 📋 PR Metadata

| Aspect | Current | Suggested |
|--------|---------|----------|
| Category | _(unlabeled)_ | `bug` |
| Milestone | _(none)_ | `0.3.11` |

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_

**Review by gnodet — COMMENTED** [review]
## Review Summary

Solid, well-motivated fix that aligns swallowed IO-error catch blocks with the project's existing conservative posture (mark affected on uncertainty). Two items worth addressing:

1. **Leftover byteman file** — `extension3/src/test/resources/byteman/fault-injection.btm` is added but never referenced by any Java test or Maven configuration. The commit history shows the byteman harness was intentionally removed (commit f1bb0c3: "remove byteman harness: the permission-flip rule adds no coverage over the direct POSIX tests"), but this .btm rule file was left behind. It should be deleted to avoid confusion.

2. **Missing test coverage for null-result path** — The behavior change in `getChangedTransitiveDependencyScope` from returning `null` to returning `"compile"` when `DependencyResolutionException.getResult()` returns null lacks test coverage. The existing `ScalpelLifecycleParticipantTest` only covers the partial-result case (module-e). A test mocking `getResult()` returning null would prove the conservative path works end-to-end.

### Strengths

- The conservative direction for all six fix sites is correct and well-justified
- Good test discipline: POSIX permission tests properly guarded for root-in-Docker environments
- Red-green commit structure makes the fix verifiable from git history
- CI passes across all matrix entries

### 📋 PR Metadata

| Aspect | Current | Suggested |
|--------|---------|----------|
| Category | _(unlabeled)_ | `bug` |
| Labels | _(none)_ | + `bug` |
| Milestone | _(none)_ | `0.4.0` |

---
_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_

**Review by gnodet — APPROVED** [review]
Clean, well-motivated fix that aligns six IO error catch blocks to the conservative direction (fail toward "affected"). The core logic is correct and the PomChangeAnalyzer tests are thorough.

**Observations (non-blocking):**
- `extension3/src/test/resources/byteman/fault-injection.btm` appears to be an orphaned file from a byteman spike (added in `4fcb46f7`, removed in `f1bb0c3`). No byteman dependency exists in any pom.xml. Consider deleting it before merge.
- The behavior change in `ScalpelLifecycleParticipant` (returning `"compile"` instead of `null` when `DependencyResolutionException` yields a null result) lacks a dedicated test. The existing test at line 221 only exercises the partial-result path. A test mocking `dependenciesResolver.resolve()` to throw with a null constructor argument would close this gap.
- Minor: in the POSIX permission tests, `Files.setPosixFilePermissions` restoration runs after the assertion — if the assertion fails, permissions aren't restored before `@TempDir` cleanup. Consider wrapping in try/finally.
- The commit structure (RED tests first, then FIX, then guard for root-in-Docker) is exemplary.

### 📋 PR Metadata

| Aspect | Current | Suggested |
|--------|---------|----------|
| Labels | _(none)_ | `bug` |
| Milestone | _(none)_ | `0.4.0` |

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_

**Comment by paoloantinori** [comment]
Rebased onto `main` @ `3fe6b72` (#145 / #131, the effective-POM-model rework). Upstream's architecture wins everywhere it changed things; this PR now grafts the #82 conservative posture onto it. Summary of what survived, what became moot, and one regression the rebase surfaced in #145 itself.

**Survived (adapted to the new architecture)**

* `PomChangeAnalyzer.checkFileForPropertyRefs`: `IOException` on a filtered resource now WARNs and returns *affected* (was silently `false`).
* `PomChangeAnalyzer.scanDirectoryForPropertyRefs`: `IOException` on `DirectoryStream` now WARNs and returns *affected* (was "skip unreadable directories").
* `PomChangeAnalyzer.isBinaryFile`: failure now WARNs (still returns non-binary so the caller proceeds to the full read, which hits the conservative path above).
* `readPomText == null` → child affected: `readPomText` is gone; the equivalent swallow was `hasEffectiveChanges` returning `false` when the child's old or new effective model could not be built. That now WARNs and returns *affected*, and `buildSingleEffectiveModel`'s total-failure logs went DEBUG → WARN.
* `getChangedTransitiveDependencyScope` null → `"compile"`: the method is gone; its failure flow reappeared in two places. First, `findChangedDependencyInTree` returned `null` when either dependency tree cannot be resolved, which also made the skip-tests safety guard treat the module as unchanged and skip its tests. Second, the propagation pass in `computeTransitivelyAffected` dropped the module on `continue`. Both now fail toward affected (WARN + conservative compile-scoped match, and a conservative transitive mark respectively). Partial results that still carry a usable graph are honored exactly as before; the existing `reportMode_includesTransitivelyAffectedModules` module-e case (partial result, empty graph, not in report) is untouched and still green.

**Became moot**

* `readPomText`: removed by #145 (text-search property fallback replaced by effective property comparison).
* `getChangedTransitiveDependencyScope`: removed (old/new dependency-tree diff via effective models).
* The Byteman spike commits: dropped entirely (they netted out to an orphaned `.btm` resource). Rationale stays as documented on the old branch: under root containers the permission flip produces no `IOException`, and Byteman cannot throw a checked `IOException` from a trigger method that does not declare it.

**New finding in #145 worth your attention: the 1 MB resource skip is non-conservative**

Pre-#145, `checkFileForPropertyRefs` took a configurable `maxResourceFileSize` (default 10 MB, `-Dscalpel.maxResourceFileSize`) and treated oversized filtered resources as *affected* with a WARN explaining the knob. #145 dropped the parameter from the call chain and hard-coded `size > 1 MB → return false`. That means a large filtered resource that references a changed property is now silently missed, an under-build swallow in exactly the territory of #82. This PR restores the conservative direction at that point (WARN + affected, with the binary check still running first so oversized binaries do not trigger rebuilds). Related: `ScalpelConfiguration.MAX_RESOURCE_FILE_SIZE` in `core` is still parsed and validated but no longer consumed by the analyzer, and the effective limit silently changed from 10 MB configurable to 1 MB hard-coded. I left the core knob in place untouched, since rewiring it through the new analyzer plumbing (or deprecating it) seemed like your call rather than mine, possibly as an #82 follow-up.

**Tests**

Six new tests, all red on `main` at the exact conservative assertion and green after the fix: unparseable child POM, unreadable filtered resource file, unreadable resource subdirectory, >1 MB filtered resource (all in `PomChangeAnalyzerTest`), and two `ScalpelLifecycleParticipantTest` report-mode tests for total resolution failure (one with valid models via the first-pass tree diff, one with missing models via the propagation pass). The old unreadable-POM test was reworked into the unparseable-POM test since `readPomText` no longer exists. POSIX permission tests keep the root-in-Docker `assumeTrue` guards.

One testing quirk we hit: on Maven 3.9.x `new DependencyResolutionException(null, ...)` NPEs inside its own constructor (`prepareDetailMessage` dereferences the result), so a "no partial result at all" failure is only representable in tests as a partial result whose `getDependencyGraph()` is null. That is the shape the new tests use, and it exercises the same null-graph guard the production code checks.

One judgment call to flag: resolution-failure marks in the transitive path reuse `REASON_TRANSITIVE_DEPENDENCY` (the same report outcome as the pre-rebase branch's `"compile"` fallback). If you would prefer a distinct reason string (e.g. `TRANSITIVE_DEPENDENCY_UNRESOLVED`), happy to add it.

**Review by gnodet — APPROVED** [review]
**Re-review (new commit `a04a082d`):** The follow-up commit correctly guards POSIX permission tests with JUnit 5 `assumeTrue` to skip them in root-in-Docker environments where permission restrictions don't apply. The change is minimal and focused.

The `assumeTrue` guards are well-placed: they sit immediately after `setPosixFilePermissions`, verifying whether the permission change actually took effect before proceeding. On non-POSIX systems or privileged processes, the test is correctly skipped rather than passing vacuously.

The previous review's try/finally concern is addressed in practice: the restoration-before-assertion order ensures assertion failures don't leave dirty permissions, and the only remaining risk (exception from `analyzeChanges`) is negligible given the conservative error handling introduced in the same PR.

All 12 CI matrix entries pass. Labels now correctly include `bug` — thank you.

### 📋 PR Metadata

| Aspect | Current | Suggested |
|--------|---------|----------|
| Milestone | _(none)_ | `0.4.0` |

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_

**Review by gnodet — APPROVED** [review]
Re-review after rebase: the 3-commit structure (RED tests → fix → POSIX guards) is clean and well-organized. All prior review concerns have been addressed — the orphaned byteman file is gone, test coverage is comprehensive across all six fix sites, and CI passes on all 12 matrix entries. The conservative error-handling direction (fail toward "affected") is correct throughout.

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_

**Comment by gnodet** [comment]
Re: the judgment call at the end — I'd go with the distinct reason string (`TRANSITIVE_DEPENDENCY_UNRESOLVED`).

The whole point of #82 is making failures visible instead of silently swallowing them. Reusing `REASON_TRANSITIVE_DEPENDENCY` is a mild form of information-swallowing — a user looking at the report can't distinguish "a transitive dep genuinely changed" from "we couldn't resolve your graph and had to be conservative." The WARN in the log helps, but the report is the primary artifact people actually look at.

The codebase already follows this pattern of fine-grained reason variants (`TRANSITIVE_DEPENDENCY` vs `TRANSITIVE_DEPENDENCY_TEST`, `UPSTREAM_DEPENDENCY` vs `DOWNSTREAM_DEPENDENT`), so a separate `TRANSITIVE_DEPENDENCY_UNRESOLVED` is consistent. And it's actionable — if a user sees it in their report, they know they have a resolution problem to investigate (missing repo, expired snapshot, network issue) rather than just an unexplained rebuild.

**Comment by paoloantinori** [comment]
Implementation of gnodet's review feedback committed as `dd53cb2`.

**Changes made:**
1. Added REASON_TRANSITIVE_DEPENDENCY_UNRESOLVED constant to ScalpelReport
2. Updated extension3 code to emit the new reason for resolution-failure marks:
   - First pass: `findChangedDependencyInTree` returning the (unresolved) sentinel
   - Propagation pass: when `resolveProjectDependencies` fails while affected modules exist
3. Added TRANSITIVE_DEPENDENCY_UNRESOLVED to the schema reasons enum
4. Extended the two resolution-failure tests to assert the new reason appears and the old one does not
5. Added row to README reason table

**Rationale:** Distinct reason strings let report consumers distinguish "a transitive dep genuinely changed" from "we couldn't resolve your graph and had to be conservative." The WARN in the log helps, but the report is the primary artifact.

Both test suites pass (core and extension3).

**Review by gnodet — APPROVED** [review]
Re-review after latest commit (dd53cb2): cleanly implements the feedback to introduce a distinct `TRANSITIVE_DEPENDENCY_UNRESOLVED` reason string, consistently applied across the constant definition, JSON schema, README, code paths, and test assertions. The overall PR remains well-structured and correct.

Two low-severity observations (non-blocking):

1. **Evidence string inconsistency** — when `depMatch.ga` is the `UNRESOLVED_GA` sentinel, the evidence becomes `"dependency tree diff (unresolved)"`, leaking the sentinel value into user-visible explain output. The propagation loop uses the cleaner `"dependency resolution failed; conservatively treated as affected"`. Consider guarding for consistency.

2. **Schema merge conflict fix** — this PR resolves merge conflict markers that currently exist on main in `scalpel-report-v2.schema.json`. The resolution is correct (keeps both `skippedModules` and `status/reason` fields) — a welcome side-effect.

The commit structure across the full PR is exemplary: red test first → green fix → environment guard → review feedback. Each commit is atomic and reviewable independently.

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_

**Comment by gnodet** [comment]
@copilot resolve the merge conflicts in this pull request

**Comment by paoloantinori** [comment]
Rebased onto current main (72d52af) and resolved the conflict in `PomChangeAnalyzer` ourselves, beating the delegated bot to it; branch force-pushed.

Since our last rebase, main gained #130 (the symlink-safe `walkFileTree` rewrite of the resource scan, which is our own prior work) and #135 (fetch validation). The resolution keeps #130's merged shape exactly as it landed: `walkFileTree` without `FOLLOW_LINKS`, the 10,000-entry visit budget, depth truncation treated conservatively, and the unreadable-directory WARN-plus-affected path. On top of that baseline only the #82 changes are grafted:

- `checkFileForPropertyRefs`: `IOException` now WARNs and returns true (affected) instead of silently returning false, and a filtered resource over the 1 MB scan limit now WARNs and returns true instead of being skipped.
- `isBinaryFile`: logs when the binary probe cannot read the file, so the caller proceeds to the full read that triggers the conservative path.
- `hasEffectiveChanges`: a child is marked affected when either effective model cannot be built.
- `buildSingleEffectiveModel`: model-build failures are logged at WARN instead of DEBUG.

The only hunk from this branch that did not survive is the old `DirectoryStream` catch, which #130's walk-level `IOException` catch supersedes with the same conservative posture.

Verification: `./mvnw -am -pl extension3 test` 171 passed, `./mvnw -pl core test` 149 passed.

**Review by gnodet — APPROVED** [review]
Re-review after latest rebase: the conservative error-handling fix remains correct across all 9 catch-block sites. The `TRANSITIVE_DEPENDENCY_UNRESOLVED` reason string cleanly implements the requested feedback.

Observations (non-blocking):

1. **Evidence string inconsistency** (low, carried from prior review) — when `depMatch.ga` is the `UNRESOLVED_GA` sentinel, the evidence becomes `"dependency tree diff (unresolved)"`, leaking the sentinel into user-visible explain output. The propagation loop uses the cleaner `"dependency resolution failed; conservatively treated as affected"`. Consider guarding for consistency.

2. **Rebase ancestry** (low) — the branch was rebased onto a commit that includes unmerged schema additions from PRs #91 and #93 (`skippedModules`, `evidence` fields). Merging will introduce those additions into main as collateral. If those features are intentionally bundled, no action needed; if not, rebase onto actual main to isolate only the #82 changes.

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_

File: extension3/src/main/java/eu/maveniverse/maven/scalpel/extension3/internal/ScalpelLifecycleParticipant.java [hunk 1]
```diff
@@ -1059,14 +1083,19 @@ private ChangedDependencyMatch findChangedDependencyInTree(
...
+            return new ChangedDependencyMatch(UNRESOLVED_GA, "compile");
         }
 
         // Resolve old dependency tree from old effective model
         DependencyResolutionResult oldResult =
                 resolveModelDependencies(oldEffectiveModel, project, session, oldCollectCache);
         if (oldResult == null || oldResult.getDependencyGraph() == null) {
-            return null;
+            // Conservative: same posture when the old tree cannot be resolved.
+            return new ChangedDependencyMatch(UNRESOLVED_GA, "compile");
```
[the remark this answers is not in the document]
Comment (gnodet) [reply extension3/src/main/java/eu/maveniverse/maven/scalpel/extension3/internal/ScalpelLifecycleParticipant.java:1098]: It should be fixed in [0ca093a](https://github.com/maveniverse/scalpel/pull/129/commits/0ca093a5ec10671e3a8fe2ad4d7cc62bd8faca97)

**Review by gnodet — APPROVED** [review]
**Re-review after new commits.** The three new commits since the last approval are all clean and appropriate — a merge from main, a necessary schema drift guard update, and a test correctness fix addressing CodeRabbit feedback. No new issues introduced.

**New commit details:**
- `23f49a5` correctly adds `TRANSITIVE_DEPENDENCY_UNRESOLVED` to the schema drift guard test, keeping Java constants and JSON schema enum in sync.
- `0ca093a` correctly narrows the POM version replacement from `<version>1.0</version>` (which matched both parent and dependency version) to `commons-lang</artifactId><version>1.0</version>` (only matches the dependency) — addressing CodeRabbit feedback.
- Merge from main is clean and brings in schema test infrastructure needed for the drift guard test.

### 📋 PR Metadata

| Aspect | Current | Suggested |
|--------|---------|----------|
| Milestone | _(none)_ | `0.4.0` |

Labels look correct (`bug`, `documentation`, `java`).

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_
